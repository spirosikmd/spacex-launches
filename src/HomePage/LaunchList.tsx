import React from 'react';
import PropTypes from 'prop-types';
import {
  withStyles,
  Theme,
  WithStyles,
  createStyles,
} from '@material-ui/core/styles';
import Launch, { LaunchPropType } from './Launch';
import Image from '../Image';
import emptyListImage from './emptyList.png';
import { LaunchData } from '../api';

const styles = (theme: Theme) =>
  createStyles({
    launchList: {
      display: 'flex',
      flexDirection: 'column' as 'column',
      position: 'relative' as 'relative',
    },
    timeline: {
      borderLeft: `1px solid ${theme.palette.grey[300]}`,
      position: 'absolute' as 'absolute',
      top: '6px',
      bottom: '6px',
      left: `${theme.spacing.unit + 4}px`,
      '&::before': {
        position: 'absolute' as 'absolute',
        content: '""',
        borderLeft: `1px dashed ${theme.palette.grey[300]}`,
        top: '-13px',
        left: '-1px',
        height: '12px',
      },
      '&::after': {
        position: 'absolute' as 'absolute',
        content: '""',
        borderLeft: `1px dashed ${theme.palette.grey[300]}`,
        bottom: '-13px',
        left: '-1px',
        height: '12px',
      },
    },
    launchItem: {
      padding: `${theme.spacing.unit * 2}px 0`,
    },
  });

interface LaunchListProps extends WithStyles<typeof styles> {
  launches: LaunchData[];
}

const LaunchList = ({ classes, launches }: LaunchListProps) => {
  if (launches.length === 0) {
    return <Image src={emptyListImage} alt="No launches..." />;
  }

  return (
    <div className={classes.launchList}>
      <div className={classes.timeline} />
      {launches.map(launch => (
        <div key={launch.flightNumber} className={classes.launchItem}>
          <Launch launch={launch} />
        </div>
      ))}
    </div>
  );
};

LaunchList.propTypes = {
  classes: PropTypes.object.isRequired,
  launches: PropTypes.arrayOf(PropTypes.shape(LaunchPropType)).isRequired,
};

export default withStyles(styles)(React.memo(LaunchList));
