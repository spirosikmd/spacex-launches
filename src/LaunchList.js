import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Launch, { LaunchPropType } from './Launch';
import emptyListImage from './emptyList.png';

const styles = theme => ({
  launchList: {
    display: 'flex',
    flexDirection: 'column',
    position: 'relative',
  },
  timeline: {
    borderLeft: `1px solid ${theme.palette.grey[300]}`,
    position: 'absolute',
    top: '6px',
    bottom: '6px',
    left: `${theme.spacing.unit + 4}px`,
    '&::before': {
      position: 'absolute',
      content: '""',
      borderLeft: `1px dashed ${theme.palette.grey[300]}`,
      top: '-13px',
      left: '-1px',
      height: '12px',
    },
    '&::after': {
      position: 'absolute',
      content: '""',
      borderLeft: `1px dashed ${theme.palette.grey[300]}`,
      bottom: '-13px',
      left: '-1px',
      height: '12px',
    },
  },
  launchItem: {
    padding: `${theme.spacing.unit * 2}px 0`,
  },
  emptyListImageWrapper: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  emptyListImage: {
    width: '400px',
    [theme.breakpoints.down('xs')]: {
      width: '300px',
    },
  },
});

export const LaunchList = ({ classes, launches }) => {
  if (launches.length === 0) {
    return (
      <div className={classes.emptyListImageWrapper}>
        <img
          src={emptyListImage}
          alt="No launches..."
          className={classes.emptyListImage}
        />
      </div>
    );
  }

  return (
    <div className={classes.launchList}>
      <div className={classes.timeline} />
      {launches.map(launch => (
        <div key={launch.flightNumber} className={classes.launchItem}>
          <Launch launch={launch} />
        </div>
      ))}
    </div>
  );
};

LaunchList.propTypes = {
  classes: PropTypes.object.isRequired,
  launches: PropTypes.arrayOf(PropTypes.shape(LaunchPropType)).isRequired,
};

export default withStyles(styles)(React.memo(LaunchList));
