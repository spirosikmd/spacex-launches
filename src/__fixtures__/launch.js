export function createLaunchResponse(data) {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data) {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
