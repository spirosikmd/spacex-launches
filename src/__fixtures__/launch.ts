import { LaunchResponseData, LaunchData } from '../api';

export function createLaunchResponse(
  data: LaunchResponseData
): LaunchResponseData {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data: LaunchData): LaunchData {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
