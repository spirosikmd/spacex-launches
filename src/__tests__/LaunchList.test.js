import { LaunchList } from '../LaunchList';
import { createLaunch } from '../__fixtures__/launch';

describe('LaunchList', () => {
  let props;

  beforeEach(() => {
    props = {
      launches: [],
    };
  });

  describe('when there are no launches', () => {
    it('renders a message', () => {
      expect(shallow(LaunchList, props)).toMatchSnapshot();
    });
  });

  describe('when there are launches', () => {
    beforeEach(() => {
      props.launches = [createLaunch(), createLaunch({ flightNumber: 2 })];
    });

    it('renders a list of launches', () => {
      expect(shallow(LaunchList, props)).toMatchSnapshot();
    });
  });
});
