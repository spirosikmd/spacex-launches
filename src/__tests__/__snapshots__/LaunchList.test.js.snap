// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchList when there are launches renders a list of launches 1`] = `
<div
  className="launchList"
>
  <div
    className="timeline"
  />
  <div
    className="launchItem"
    key="1"
  >
    <WithStyles(Component)
      launch={
        Object {
          "details": "Details",
          "flightNumber": 1,
          "isFailed": false,
          "isSuccessful": true,
          "isTentative": false,
          "isUpcoming": false,
          "missionIds": Array [
            "id1",
          ],
          "missionName": "Test",
          "utcDate": Object {
            "getDate": [MockFunction],
            "getFullYear": [MockFunction],
            "getMonth": [MockFunction],
            "toLocaleString": [MockFunction],
          },
        }
      }
    />
  </div>
  <div
    className="launchItem"
    key="2"
  >
    <WithStyles(Component)
      launch={
        Object {
          "details": "Details",
          "flightNumber": 2,
          "isFailed": false,
          "isSuccessful": true,
          "isTentative": false,
          "isUpcoming": false,
          "missionIds": Array [
            "id1",
          ],
          "missionName": "Test",
          "utcDate": Object {
            "getDate": [MockFunction],
            "getFullYear": [MockFunction],
            "getMonth": [MockFunction],
            "toLocaleString": [MockFunction],
          },
        }
      }
    />
  </div>
</div>
`;

exports[`LaunchList when there are no launches renders a message 1`] = `
<div>
  <img
    alt="No launches..."
    src="emptyList.webp"
  />
</div>
`;
